File: src/main/java/com/gtladd/gtladditions/common/items/behavior/AstralArrayBehavior.java
package com.gtladd.gtladditions.common.items.behavior;

import com.gregtechceu.gtceu.api.blockentity.MetaMachineBlockEntity;
import com.gregtechceu.gtceu.api.item.component.IInteractionItem;

import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.item.context.UseOnContext;

import com.gtladd.gtladditions.api.machine.IAstralArrayInteractionMachine;
import org.jetbrains.annotations.NotNull;

public class AstralArrayBehavior implements IInteractionItem {

    public static final AstralArrayBehavior INSTANCE = new AstralArrayBehavior();

    private AstralArrayBehavior() {}

    @Override
    public @NotNull InteractionResult useOn(UseOnContext context) {
        if (context.getLevel().getBlockEntity(context.getClickedPos()) instanceof MetaMachineBlockEntity machineBlock) {
            if (machineBlock.getMetaMachine() instanceof IAstralArrayInteractionMachine machine) {
                if (context.getPlayer() instanceof ServerPlayer serverPlayer && serverPlayer.isShiftKeyDown()) {
                    int used = machine.increaseAstralArrayCount(context.getItemInHand().getCount());

                    if (used > 0) {
                        if (!serverPlayer.isCreative()) {
                            context.getItemInHand().shrink(used);
                        }

                        serverPlayer.sendSystemMessage(
                                Component.translatable("gtladditions.message.astral_array.installed_multiple",
                                        used, machine.getAstralArrayCount()));
                    } else {
                        serverPlayer.sendSystemMessage(
                                Component.translatable("gtladditions.message.astral_array.max_reached"));
                    }

                    return InteractionResult.SUCCESS;
                }
            }
        }
        return InteractionResult.PASS;
    }
}

File: src/main/java/com/gtladd/gtladditions/common/items/behavior/AstralArrayBehavior.kt
package com.gtladd.gtladditions.common.items.behavior

import com.gregtechceu.gtceu.api.blockentity.MetaMachineBlockEntity
import com.gregtechceu.gtceu.api.item.component.IInteractionItem
import com.gtladd.gtladditions.api.machine.IAstralArrayInteractionMachine
import com.gtladd.gtladditions.utils.ComponentExtensions.toComponent
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.InteractionResult
import net.minecraft.world.item.context.UseOnContext

object AstralArrayBehavior : IInteractionItem {

    override fun useOn(context: UseOnContext): InteractionResult {
        val player = context.player ?: return InteractionResult.PASS
        if (!player.isShiftKeyDown) return InteractionResult.PASS

        val machine = (context.level.getBlockEntity(context.clickedPos) as? MetaMachineBlockEntity)
            ?.metaMachine as? IAstralArrayInteractionMachine
            ?: return InteractionResult.PASS

        if (context.level.isClientSide) return InteractionResult.SUCCESS

        val serverPlayer = player as ServerPlayer
        val equivalent = IAstralArrayInteractionMachine.getAstralArrayEquivalent(context.itemInHand)
        if (equivalent <= 0) return InteractionResult.PASS
        val itemCount = context.itemInHand.count
        val installedCount = machine.increaseAstralArrayCount(itemCount * equivalent)

        when {
            installedCount > 0 -> {
                if (!serverPlayer.isCreative) {
                    context.itemInHand.shrink((installedCount + equivalent - 1) / equivalent)
                }

                serverPlayer.sendSystemMessage(
                    "gtladditions.message.astral_array.installed_multiple".toComponent(installedCount, machine.astralArrayCount)
                )
            }
            else -> {
                serverPlayer.sendSystemMessage(
                    "gtladditions.message.astral_array.max_reached".toComponent
                )
            }
        }

        return InteractionResult.SUCCESS
    }
}